File: Mirage/Mirage Wallpaper/SteamSetup/SteamSetupViewModel.swift
//
//  Mirage Wallpaper
//
//  Copyright © 2026 王孝慈. All rights reserved.
//

import SwiftUI

class SteamSetupViewModel: ObservableObject {
    @Published var currentStep: Int = 0
    @Published var steamCMDPath: URL?
    @Published var steamCMDInstallState: SteamCMDInstallState = .detecting
    @Published var loginState: SteamLoginState = .idle

    @Published var username: String = ""
    @Published var password: String = ""
    @Published var guardCode: String = ""
    @Published var errorMessage: String?

    @Published var loginLog: [String] = []

    let totalSteps = 4

    var canProceed: Bool {
        switch currentStep {
        case 0:
            return true
        case 1:
            switch steamCMDInstallState {
            case .found, .installed: return true
            default: return false
            }
        case 2:
            return loginState == .success
        case 3:
            return true
        default:
            return false
        }
    }

    init() {
        username = SteamCMDManager.shared.savedUsername
    }

    func detectSteamCMD() {
        steamCMDInstallState = .detecting
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            let found = SteamCMDManager.shared.detectSteamCMD()
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                if let path = found {
                    self?.steamCMDPath = path
                    self?.steamCMDInstallState = .found(path.path)
                } else {
                    self?.steamCMDInstallState = .notFound
                }
            }
        }
    }

    func installSteamCMD() {
        SteamCMDManager.shared.installSteamCMD { [weak self] state in
            self?.steamCMDInstallState = state
            if case .installed(let path) = state {
                self?.steamCMDPath = URL(fileURLWithPath: path)
            }
        }
    }

    func cancelSteamCMDInstallation() {
        SteamCMDManager.shared.cancelSteamCMDInstallation()
    }

    func login() {
        guard !username.isEmpty, !password.isEmpty else {
            errorMessage = "请输入用户名和密码"
            return
        }
        errorMessage = nil
        loginLog.removeAll()
        SteamCMDManager.shared.login(
            username: username,
            password: password,
            onLog: { [weak self] line in
                self?.loginLog.append(line)
            }
        ) { [weak self] state in
            self?.loginState = state
            if case .failed(let msg) = state {
                self?.errorMessage = msg
            }
            if case .success = state {
                self?.password = ""
                self?.guardCode = ""
            }
            if case .waitingForGuard = state {
                self?.errorMessage = nil
            }
        }
    }

    func submitGuardCode() {
        guard !guardCode.isEmpty else {
            errorMessage = "请输入验证码"
            return
        }
        guard SteamCMDManager.shared.submitGuardCode(guardCode) else {
            errorMessage = "Steam Guard 会话已结束，请重新登录"
            loginState = .failed("Steam Guard 会话已结束")
            return
        }
        guardCode = ""
        errorMessage = nil
        loginState = .loggingIn
    }

    func cancelLogin() {
        SteamCMDManager.shared.cancelLogin()
        password = ""
        guardCode = ""
        loginState = .idle
        errorMessage = nil
    }

    func completeSetup() {
        SteamCMDManager.shared.savedUsername = username
    }

    func nextStep() {
        guard currentStep < totalSteps - 1 else { return }
        currentStep += 1
    }

    func previousStep() {
        guard currentStep > 0 else { return }
        currentStep -= 1
    }
}

File: MirageQt/Sources/Services/SteamSetupViewModel.cpp
#include "Services/SteamSetupViewModel.h"

#include "Services/SteamCMDManager.h"

#include <QClipboard>
#include <QGuiApplication>

namespace Mirage {

namespace {

QString steamInstallStateKey(SteamCMDInstallState state) {
    switch (state) {
    case SteamCMDInstallState::Detecting: return QStringLiteral("detecting");
    case SteamCMDInstallState::Found: return QStringLiteral("found");
    case SteamCMDInstallState::NotFound: return QStringLiteral("notFound");
    case SteamCMDInstallState::Downloading: return QStringLiteral("downloading");
    case SteamCMDInstallState::Extracting: return QStringLiteral("extracting");
    case SteamCMDInstallState::Initializing: return QStringLiteral("initializing");
    case SteamCMDInstallState::Installed: return QStringLiteral("installed");
    case SteamCMDInstallState::Failed: return QStringLiteral("failed");
    }
    return {};
}

QString steamLoginStateKey(SteamLoginState state) {
    switch (state) {
    case SteamLoginState::Idle: return QStringLiteral("idle");
    case SteamLoginState::LoggingIn: return QStringLiteral("loggingIn");
    case SteamLoginState::WaitingForGuard: return QStringLiteral("waitingForGuard");
    case SteamLoginState::Success: return QStringLiteral("success");
    case SteamLoginState::Failed: return QStringLiteral("failed");
    }
    return {};
}

QString steamGuardTypeKey(SteamGuardType type) {
    switch (type) {
    case SteamGuardType::None: return QStringLiteral("");
    case SteamGuardType::Email: return QStringLiteral("email");
    case SteamGuardType::Mobile: return QStringLiteral("mobile");
    case SteamGuardType::MobileConfirm: return QStringLiteral("mobileConfirm");
    }
    return {};
}

} // namespace

SteamSetupViewModel::SteamSetupViewModel(SteamCMDManager* cmd, QObject* parent)
    : QObject(parent)
    , m_cmd(cmd) {
    connect(m_cmd, &SteamCMDManager::installStateChanged, this,
            [this](SteamCMDInstallState state, double progress, const QString& message) {
                m_installState = steamInstallStateKey(state);
                m_installProgress = progress;
                m_installMessage = message;
                emit steamChanged();
            });
    connect(m_cmd, &SteamCMDManager::loginStateChanged, this,
            [this](SteamLoginState state, const QString& message) {
                m_loginState = steamLoginStateKey(state);
                m_loginMessage = message;
                emit steamChanged();
            });
    connect(m_cmd, &SteamCMDManager::guardTypeChanged, this,
            [this](SteamGuardType) { emit steamChanged(); });
    connect(m_cmd, &SteamCMDManager::diagnosticEvent, this,
            [this](const QString& line) {
                m_loginLog.append(line);
                if (m_loginLog.size() > 500)
                    m_loginLog.remove(0, m_loginLog.size() - 500);
                emit steamChanged();
            });
    connect(m_cmd, &SteamCMDManager::steamCMDPathChanged, this,
            [this](const QString&) { emit steamChanged(); });
    connect(m_cmd, &SteamCMDManager::authenticationChanged, this,
            [this](bool, const QString&) { emit steamChanged(); });
    connect(m_cmd, &SteamCMDManager::sessionReusableChanged, this,
            [this](bool) { emit steamChanged(); });
}

QString SteamSetupViewModel::installState() const { return m_installState; }
double SteamSetupViewModel::installProgress() const { return m_installProgress; }
QString SteamSetupViewModel::installMessage() const { return m_installMessage; }
QString SteamSetupViewModel::loginState() const { return m_loginState; }
QString SteamSetupViewModel::loginMessage() const { return m_loginMessage; }
QStringList SteamSetupViewModel::loginLog() const { return m_loginLog; }
QString SteamSetupViewModel::guardType() const { return steamGuardTypeKey(m_cmd->steamGuardType()); }
bool SteamSetupViewModel::sessionReusable() const { return m_cmd->sessionReusable(); }

void SteamSetupViewModel::detect() {
    m_installState = QStringLiteral("detecting");
    emit steamChanged();
    m_cmd->detectSteamCMD();
}

void SteamSetupViewModel::install() {
    m_cmd->installSteamCMD();
}

void SteamSetupViewModel::cancelInstallation() {
    m_cmd->cancelInstallation();
}

void SteamSetupViewModel::login(const QString& username, const QString& password) {
    m_loginLog.clear();
    emit steamChanged();
    m_cmd->login(username, password);
}

void SteamSetupViewModel::submitGuardCode(const QString& code) {
    m_cmd->submitGuardCode(code);
}

void SteamSetupViewModel::confirmMobileLogin() {
    m_cmd->confirmMobileLogin();
}

void SteamSetupViewModel::useSavedSession() {
    m_loginState = QStringLiteral("loggingIn");
    m_loginMessage = QStringLiteral("正在验证已保存的 SteamCMD 会话");
    emit steamChanged();
    m_cmd->refreshSession();
}

void SteamSetupViewModel::cancelLogin() {
    m_cmd->cancelLogin();
}

void SteamSetupViewModel::cancelPendingWork() {
    m_cmd->cancelLogin();
    m_cmd->cancelInstallation();
}

void SteamSetupViewModel::logout() {
    m_cmd->logout();
    m_loginState = QStringLiteral("idle");
    m_loginMessage = QStringLiteral("未登录");
    emit steamChanged();
}

void SteamSetupViewModel::copyLoginLog() {
    QGuiApplication::clipboard()->setText(m_loginLog.join(QStringLiteral("\n")));
}

} // namespace Mirage
